### PR TITLE
placement: add ruleList to accelerate rule search

### LINKED ISSUE #2163: Improve performance of finding placement rules of a region
### Describe your feature request related problem: 
current implementation needs to iterate all placement rules. It should be too slow when there are thouthands of rules.

### Describe the feature you'd like:
Use some preprocessing to make it faster.

### PR DESCRIPTION
**Pull request by disksing**
Signed-off-by: disksing <i@disksing.com>

### What problem does this PR solve? 
Fix #2163

### What is changed and how it works?
Pre-calculate rules for each range when any rule is updated.

### Check List 

Tests 
 - Unit test
 - Manual test (create 10000 rules then `curl pd/api/v1/config/rule/key/{key}`)

Related changes
 - Need to cherry-pick to the release branch

### CODE REVIEW

File: server/schedule/placement/rule_list.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	startKey   []byte
+	rules      []*Rule
+	applyRules []*Rule
+}
+
+type ruleList struct {
+	ranges []rangeRules // ranges[i] contains rules apply to (ranges[i].startKey, ranges[i+1].startKey).
+}
+
+func buildRuleList(rules map[[2]string]*Rule) ruleList {
```
Comment (rleungx) [inline server/schedule/placement/rule_list.go]: How about directly calling this function as `updateRuleList`?

Comment (disksing) [reply server/schedule/placement/rule_list.go]: good idea

**Comment by disksing** [comment]
ping /cc @rleungx

**Pull request merged**